contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,364 @@
+      (uint coveredAmount, uint price) = buyCoverFromPool(stakingPool, productId, amountToCover, period);
+      amountToCover -= coveredAmount;
+      totalPremiumInNXM += price;
+      stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+    }
+    require(amountToCover == 0, "Not enough available capacity");
+
+    premiumInAsset = totalPremiumInNXM * pool().getTokenPrice(pool().assets(payoutAsset)) / 1e18;
+
+    covers.push(Cover(
+        productId,
+        payoutAsset,
+        uint96(amount),
+        uint32(block.timestamp + 1),
+        uint32(period),
+        uint96(premiumInAsset)
+      ));
+
+    coverId = covers.length - 1;
+    _safeMint(msg.sender, coverId);
+  }
+
+  function buyCoverFromPool(
+    IStakingPool stakingPool,
+    uint24 productId,
+    uint amountToCover,
+    uint32 period
+  ) internal returns (uint, uint) {
+
+    uint availableCapacity = stakingPool.getAvailableCapacity(productId, capacityFactors[productId]);
+
+    uint coveredAmount;
+    if (amountToCover > availableCapacity) {
+      coveredAmount = availableCapacity;
+    } else {
+      coveredAmount = amountToCover;
+    }
+
+    uint capacityFactor = capacityFactors[productId];
+    (uint basePrice, uint price) = getPrice(coveredAmount, period, productId, stakingPool);
+    lastPrices[productId][address(stakingPool)] = basePrice;
+    lastPriceUpdate[productId][address(stakingPool)] = block.timestamp;
+
+    stakingPool.buyCover(
+      productId,
+      coveredAmount,
+      REWARD_BPS * price / BASIS_PRECISION,
+      period,
+      capacityFactor
+    );
+
+    stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+
+    return (coveredAmount, price);
+  }
+
+  function extendCover(
+    uint coverId,
+    uint32 period,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external returns (uint) {
+    require(_isApprovedOrOwner(_msgSender(), coverId), "Cover: caller is not owner nor approved");
+
+    Cover memory cover = covers[coverId];
+    (uint newCoverId, uint priceInAsset) = _createCover(
+      ERC721.ownerOf(coverId),
+      cover.productId,
+      cover.payoutAsset,
+      amount,
+      period,
+      stakingPools
+    );
+
+    // make the cover expire at current block
+    uint32 newPeriod = uint32(block.timestamp) - cover.start;
+    uint32 previousPeriod = covers[coverId].period;
+    uint priceAlreadyPaid = (previousPeriod - newPeriod) / previousPeriod * cover.price;
+    covers[coverId].period = newPeriod;
+
+    if (priceInAsset > priceAlreadyPaid) {
+      // get price for already paid asset
+      uint priceToBePaid = priceInAsset - priceAlreadyPaid;
+      require(priceToBePaid <= maxPrice, "Cover: Price exceeds maxPrice");
+      retrievePayment(priceToBePaid, cover.payoutAsset);
+    }
+
+    return newCoverId;
+  }
+
+  function addAmount(
+    uint coverId,
+    uint96 amount,
+    uint maxPrice,
```

Comment:
what is maxPrice here? we should either stick to premium % slippage, or absolute annual price % slippage